### PR TITLE
chore: Replace `poetry` with `uv`

### PR DESCRIPTION
**Pull request by aborgna-q**
Replaces the python project manager with `uv`.
For most commands that used poetry before, it just be as easy as replacing `poetry {things}` with `uv {things}`.

Some notable exeptions:

- Instead of `poetry install --with extra1,extra2`,
  use `uv sync --extra extra1 --extra extra2`.
  Note that will delete any extraneous packages/extras so the venv state is deterministic.
  Add `--inexact` if you want to replicate the old poetry behaviour.

- `poetry shell` is no longer available. The preferred workflow is to `uv run` / `uvx` things, to ensure you are always in sync with the project definition.
  If that's too annoying, it is always possible to `source .venv/bin/activate`.

Updates the justfile, so using that should be the same as before.

### CODE REVIEW

File: .github/workflows/ci-py.yml [hunk 1]
```diff
@@ -95,34 +105,63 @@ jobs:
...
-    name: test python ${{ matrix.python-version }}
+    name: test python ${{ matrix.python-version.py }}
     runs-on: ubuntu-latest
 
     strategy:
       matrix:
-        python-version: ['3.10', '3.12']
+        python-version:
+          - { py: '3.10', coverage: false }
+          - { py: '3.12', coverage: true }
```
Comment (mark-koch) [inline .github/workflows/ci-py.yml:115]: Good idea to combine the coverage run with regular testing 👍

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -36,31 +36,24 @@ repos:
 
   - repo: local
     hooks:
-      - id: poetry-install
-        name: sync poetry deps
-        description: Ensure development tools are installed.
-        entry: poetry install --sync
-        language: system
-        files: poetry.lock
-        pass_filenames: false
```
Comment (mark-koch) [inline .pre-commit-config.yaml:45]: Do we want to `uv sync` here?

File: justfile [hunk 3]
```diff
@@ -27,37 +27,32 @@ bench language="[rust|python]": (_run_lang language \
...
     )
 
 # Run unsoundness checks using miri
 miri:
     PROPTEST_DISABLE_FAILURE_PERSISTENCE=true MIRIFLAGS='-Zmiri-env-forward=PROPTEST_DISABLE_FAILURE_PERSISTENCE' cargo +nightly miri test
 
-# Load a shell with all the dependencies installed
-shell:
-    poetry shell
-
```
Comment (mark-koch) [inline justfile:52]: We could do `source .venv/bin/activate` here, but not sure if it's a good idea...

**Review by mark-koch — APPROVED** [review]
image

File: justfile (hunk 3 above)
Comment (aborgna-q) [reply justfile:52]: Doesn't really work with how justfile runs its commands :/

File: .pre-commit-config.yaml (hunk 2 above)
Comment (aborgna-q) [reply .pre-commit-config.yaml:45]: `sync` is done automatically each time you call `uv run`.
In CI we add the extra step to have more feedback and pass `-locked` as needed, but here it should be instantaneous.

**Pull request merged**